llvm/lib/Transforms/IPO/SampleProfile.cpp

```diff
@@ -1630,7 +1630,7 @@ void SampleProfileLoader::promoteMergeNotInlinedContextSamples(
         // separate map so that it does not rehash the original profile.
         if (!OutlineFS)
           OutlineFS = &OutlineFunctionSamples[
-              FunctionSamples::getCanonicalFnName(Callee->getName())];
+              FunctionId(FunctionSamples::getCanonicalFnName(Callee->getName()))];
```

Comment:
pretty much a drive-by comment

It seems to me `SampleContext` could take either a StringRef or a FunctionId in its constructor, and the former asserts names are not empty (https://github.com/llvm/llvm-project/blob/5f295552f1eb9b00efe0d83c5f8e6c9b948cb238/llvm/include/llvm/ProfileData/SampleProf.h#L529) while the latter doesn't (`FunctionId` ctor doesn't require names are not empty either); and this PR wants to fix a crash for the former constructor.

If I understand correctly, is this `FunctionId` needed here and below?

Reply:
It is intentionally using FunctionId constructor here to avoid that assert, and any other potential situation where a function with empty name is passed in. Since the front end is relatively independent, so we are not making any assumption there. Accepting an empty string in profile matching will at worst degrade the profile's accuracy but still result in a correct binary, which is something we want to ensure here: If both the source code and the sample profile file are valid, the compiler should not exit abnormally.

Also there's an upcoming patch which will make this moot (https://github.com/huangjd/llvm-project/tree/md5phase3) I am still validating it.

Reply:
Should the StringRef based constructor in SampleConext be removed to avoid future confusion?

Reply:
That will be addressed in MD5phase3 (pending), in that patch FunctionId no longer requires explicit constructor since the difference between MD5 and String FunctionId will be minimized